### PR TITLE
Prevent batch2 job execution to stop for empty chunk when last job st…

### PR DESCRIPTION
**Pull request by jmarchionatto**
…ep is a reduction.

Add output to bulk export result even when empty.

### CODE REVIEW

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/JobStepExecutor.java [hunk 1]
```diff
@@ -92,6 +93,11 @@ public void executeStep() {
 		}
 	}
 
+	private boolean lastStepIsReduction() {
+		int stepCount = myDefinition.getSteps().size();
+		return stepCount >= 1 && myDefinition.getSteps().get(stepCount - 1).isReductionStep();
```
Comment (lukedegruchy) [inline hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/JobStepExecutor.java:98]: Java Collections really need a getLast()

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/progress/JobInstanceProgressCalculator.java [hunk 2]
```diff
@@ -96,8 +104,21 @@ public InstanceProgress calculateInstanceProgress(String instanceId) {
...
+		instanceProgress.calculateNewStatus(lastStepIsReduction(instanceId));
 
 		return instanceProgress;
 	}
+
+	private boolean lastStepIsReduction(String theInstanceId) {
+		JobInstance jobInstance = getJobInstance(theInstanceId);
+		JobDefinition<IModelJson> jobDefinition = myJobDefinitionRegistry.getJobDefinitionOrThrowException(jobInstance);
+		int stepCount = jobDefinition.getSteps().size();
+		return stepCount >= 1 && jobDefinition.getSteps().get(stepCount - 1).isReductionStep();
```
Comment (lukedegruchy) [inline hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/progress/JobInstanceProgressCalculator.java:116]: Could we somehow define a common method used by this code and the JobStepExecutor?

File: hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/export/BulkDataExportProvider.java [hunk 3]
```diff
@@ -502,6 +502,9 @@ public void exportPollStatus(
 
 						String serverBase = getServerBase(theRequestDetails);
 
+						// an output is required, even if empty, according to HL7 FHIR IG
+						bulkResponseDocument.getOutput();
```
Comment (lukedegruchy) [inline hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/export/BulkDataExportProvider.java:506]: It's annoying that you have to initialize the List like this.  Is there a better way?

**Review by lukedegruchy — APPROVED** [review]

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/progress/JobInstanceProgressCalculator.java (hunk 2 above)
Comment (michaelabuckley) [reply hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/progress/JobInstanceProgressCalculator.java:116]: How about making this a method on JobDefinition?

**Review by michaelabuckley — APPROVED** [review]
Consider adding a test to BulkExportUseCaseTest for the empty result case.  There are many pieces between this and the final result that could break on empty.

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/JobStepExecutor.java (hunk 1 above)
Comment (jmarchionatto) [reply hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/JobStepExecutor.java:98]: Absolutely!

File: hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/export/BulkDataExportProvider.java (hunk 3 above)
Comment (jmarchionatto) [reply hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/export/BulkDataExportProvider.java:506]: No other method I could find for that in that API

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/progress/JobInstanceProgressCalculator.java (hunk 2 above)
Comment (jmarchionatto) [reply hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/progress/JobInstanceProgressCalculator.java]: Done.
